File: leetcode/algorithms/108_convert_sorted_array_to_binary_search_tree/main.cpp
#include <vector>
using namespace std;

class ConvertSortedArrayToBinarySearchTree {
public:
    struct TreeNode {
        int val;
        TreeNode* left;
        TreeNode* right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
    };

    /**
     * Divide-and-conquer
     *
     * Complexities:
     *   N - Size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(logᴺ)
     */
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.size() == 0) {
            return nullptr;
        }

        return convertToBST(nums, 0, nums.size() - 1);
    }

    TreeNode* convertToBST(vector<int> nums, int left, int right) {
        if (right < left) {
            return nullptr;
        }

        int mid = (left + right) / 2;
        TreeNode* node = new TreeNode(nums[mid]);
        node->left = convertToBST(nums, left, mid - 1);
        node->right = convertToBST(nums, mid + 1, right);

        return node;
    }


    // Solution
    /**
     * Divide-and-conquer
     *
     * Complexities:
     *   N - Size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(logᴺ)
     */
    TreeNode* buildBST(const std::vector<int>& nums, int left, int right) {
        if (left > right) {
            return nullptr;
        }

        int mid = left + (right - left) / 2;

        TreeNode* node = new TreeNode(nums[mid]);

        node->left = buildBST(nums, left, mid - 1);
        
        node->right = buildBST(nums, mid + 1, right);

        return node;
    }

    TreeNode* solution(std::vector<int>& nums) {
        if (nums.empty()) {
            return nullptr;
        }

        return buildBST(nums, 0, nums.size() - 1);
    }
};

File: leetcode/algorithms/108_convert_sorted_array_to_binary_search_tree/main.java
import java.util.Arrays;

class ConvertSortedArrayToBinarySearchTree {
  public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
      this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
      this.val = val;
      this.left = left;
      this.right = right;
    }
  }

  /**
   * Divide-and-conquer
   *
   * Complexities:
   *   N - Size of `nums`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(logᴺ)
   */
  public TreeNode sortedArrayToBST(int[] nums) {
    if (nums.length == 0) {
      return null;
    }

    int mid = Math.floorDiv(nums.length, 2);
    TreeNode root = new TreeNode(nums[mid]);

    root.left = sortedArrayToBST(Arrays.copyOfRange(nums, 0, mid));
    root.right = sortedArrayToBST(Arrays.copyOfRange(nums, mid + 1, nums.length));

    return root;
  }


  // Solution
  /**
   * Divide-and-conquer
   *
   * Complexities:
   *   N - Size of `nums`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(logᴺ)
   */
    public TreeNode solution(int[] nums) {
        if (nums == null || nums.length == 0) {
            return null;
        }

        return constructBST(nums, 0, nums.length - 1);
    }

    private TreeNode constructBST(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }

        int mid = left + (right - left) / 2;

        TreeNode node = new TreeNode(nums[mid]);

        node.left = constructBST(nums, left, mid - 1);
        node.right = constructBST(nums, mid + 1, right);

        return node;
    }
}
